From e970db9e66ab7d3177dd70bed92cd742bad57824 Mon Sep 17 00:00:00 2001
From: David Heidelberg <david@ixit.cz>
Date: Fri, 21 Nov 2025 20:13:18 +0100
Subject: [PATCH 016/211] Qualcomm 3rd gen fuel gauge support
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This FG (fuel gauge) can be found on Snapdragon 835, 845, 660, 670..

Series enables the FG, which is used on phones as OnePlus 6, 6T, Pixel 3a..

Since many downstream trees (sdm845-mainline, sdm660, sdm670) carring
these patches, after Barnabás dropped older gen support from the driver,
cleaned up the driver, I picked changes and cleaned up device-tree
patches and sending the whole bulk for review.

Note:
Usually I picked the oldest commits authors.
Since there was multiple patches, multiple names and modifications etc.,
if someone from the authors, co-developers, contributors etc. missing,
please let me know and I'll fix this up in the next series.

Currently some authors are unreachable or haven't given Signed-off-by
yet, thus RFC, when this get cleaned up, I'll sent non-RFC or leave it
to the authors to continue upstreaming process in case they want to.

Until merged (or taken over by the original authors), the series is
available at:
  https://gitlab.com/sdm845/sdm845-next/-/commits/b4%2Fpmi8998_fuel_gauge
# Describe the purpose of this series. The information you put here
# will be used by the project maintainer to make a decision whether
# your patches should be reviewed, and in what priority order. Please be
# very detailed and link to any relevant discussions or sites that the
# maintainer can review to better understand your proposed changes. If you
# only have a single patch in your series, the contents of the cover
# letter will be appended to the "under-the-cut" portion of the patch.

# Lines starting with # will be removed from the cover letter. You can
# use them to add notes or reminders to yourself. If you want to use
# markdown headers in your cover letter, start the line with ">#".

# You can add trailers to the cover letter. Any email addresses found in
# these trailers will be added to the addresses specified/generated
# during the b4 send stage. You can also run "b4 prep --auto-to-cc" to
# auto-populate the To: and Cc: trailers based on the code being
# modified.

To: Sebastian Reichel <sre@kernel.org>
To: Rob Herring <robh@kernel.org>
To: Krzysztof Kozlowski <krzk+dt@kernel.org>
To: Conor Dooley <conor+dt@kernel.org>
To: Casey Connolly <casey.connolly@linaro.org>
To: Casey Connolly <casey@connolly.tech>
To: Joel Selvaraj <foss@joelselvaraj.com>
To: Yassine Oudjana <y.oudjana@protonmail.com>
To: Bjorn Andersson <andersson@kernel.org>
To: Konrad Dybcio <konradybcio@kernel.org>
To: Alexander Martinz <amartinz@shiftphones.com>
To: Barnabás Czémán <barnabas.czeman@mainlining.org>
To: Richard Acayan <mailingradian@gmail.com>
To: Alexey Minnekhanov <alexeymin@postmarketos.org>
To: Lee Jones <lee@kernel.org>
To: Stephen Boyd <sboyd@kernel.org>
Cc: linux-pm@vger.kernel.org
Cc: devicetree@vger.kernel.org
Cc: linux-kernel@vger.kernel.org
Cc: linux-arm-msm@vger.kernel.org
Cc: phone-devel@vger.kernel.org
Signed-off-by: David Heidelberg <david@ixit.cz>
---
Changes in v2:
- Connect power/supply/qcom,pmi8998-fg.yaml into mfd/qcom,spmi-pmic.yaml
  (David)
- Introduce qcom,pm660-fg compatible. (Dmitry)
- Documentation fixes. (Dmitry)
- Dropped debugging statment which can be provided by regmap. (Dmitry)
- Refactor control flow to use early exits and remove unnecessary else
  blocks in pmi8998_fg_get_property. (Dmitry)
- Return early in pmi8998_fg_get_property. (Dmitry)
- Document IMA exception bits. (Dmitry)
- Moved DMA processing to own function pmi8998_fg_clear_dma(). (Dmitry)
- Defined multiple values.
- Refactor control flow for chg_psy to early exit and remove unnecessary conditions.
- Link to v1: https://lore.kernel.org/r/20251124-pmi8998_fuel_gauge-v1-0-dd3791f61478@ixit.cz

Changes in v1:
- Clean and read charge full and max voltage from dts. (Joel)
- Dont put battery info on remove. All the battery info data is devm
  tracked so it will be freed for us on removal. (Casey)
- Invert charging current. Battery current should be negative for charging
  batteries, invert it so it's correct. Fixes upower charging status
  reporting. (Casey)
- Expose PROP_STATUS to fix upower not detecting charging status properly.
- Adopt battery info API changes.
- Silence -EPROBE_DEFER error. (Richard)
- Add present sysfs property. Required for battery drivers from
  UPower>=1.90.0 onwards to work properly. (Alexander)
- Drop support for older gen from the driver. (Barnabás)
- Rename the driver to pmi8998_fg. (Barnabás)
- Remove unnecessary batt_info. (Barnabás)
- Fix all checkpatch issues in strict mode. (Barnabás)
- Fixed naming, properties sorting, styling issues, few checkpatch
  warnings. (David)
- Updated Joel to "Joel Selvaraj <foss@joelselvaraj.com>" everywhere
  (patch author, author field inside the driver, dt docs). (David)
- Link to v0:
  - https://gitlab.com/msm8998-mainline/linux/ (very old)
  - https://gitlab.com/sdm845-mainline/linux/ (multiple branches)
  - https://github.com/sdm660-mainline/linux/
  - https://gitlab.com/sdm670-mainline/linux-patches/-/tree/main/stable

--- b4-submit-tracking ---
# This section is used internally by b4 prep for tracking purposes.
{
  "series": {
    "revision": 2,
    "change-id": "20251121-pmi8998_fuel_gauge-cfaf78ed103d",
    "prefixes": [
      "RFC"
    ],
    "history": {
      "v1": [
        "20251124-pmi8998_fuel_gauge-v1-0-dd3791f61478@ixit.cz"
      ]
    },
    "prerequisites": []
  }
}
-- 
2.55.0

